### PR TITLE
Add developer docs for new features in IoT Edge

### PR DESCRIPTION
**Pull request by shantanu1singh**
Add developer docs for new features in IoT Edge:

* Module processing priority
* Edge Hub database store size limiting
* System modules in-memory database store backup and restore

### CODE REVIEW

File: doc/ModulePriority.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
+# How to configure priorities for modules
+
+By default, IoT Edge does not impose an ordering in the sequence in which modules are started, updated or stopped. Edge Agent by default is the first module that gets started and based on the edge deployment specification, it figures out which modules need to be started, updated or stopped and executes those operations in a non-deterministic order.
+
+The processing order of modules can be controlled by specifying the value of a module-specific property called `priority` in the IoT Edge deployment. Modules that have been assigned a higher priority will be processed before modules that have been assigned a lower priority.
+
+**It must be noted that this feature is currently only supported for the in-memory database store of Edge Hub i.e., when the environment variable `UsePersistentStorage` is set to `false`. The work to support this feature for the disk-based database store is currently in progress.**
```
Comment (kgremban) [inline doc/ModulePriority.md]: Is this note a carry-over from the other two articles? It doesn't seem relevant to module priority.

File: doc/ModulePriority.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+As an example, some customers want the Edge Hub module to be started before any other non-system modules in the ecosystem are started. This is so that other modules don't spend unnecessary cycles waiting for Edge Hub to come up before they can start sending messages to other modules or upstream to IoT Hub.
+
+**That being said, module owners should design their modules to withstand any failures of these 'critical' or 'foundation' modules, that they are dependent upon, as they could go down at any arbitrary time and an arbitrary number of times.**
+
+## __Configuration__
+
+Customers can optionally specify a `priority` value for each module in their IoT Edge deployment to achieve module boot ordering. Modules with a higher priority will be created and started first and only when an attempt has been made to start them, will other lower priority modules be created and started.
+
+The value of `priority` will be positive and zero-based with 0 being the highest priority value.
+A higher numeric `priority` value would indicate a lower priority assigned to that module. The maximum value of this property will be 4294967295.
```
Comment (kgremban) [inline doc/ModulePriority.md]: This is product feedback, not doc feedback, but this system of prioritization is backwards from the way IoT Edge prioritizes automatic deployments. It might cause confusion for customers when 1 is the lowest priority for deployments, but the highest priority for module within the deployment.

File: doc/ModulePriority.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+**It must be noted that module boot ordering will initially not be supported for the Kubernetes execution mode of IoT Edge.**
+
+## __Example__
+
+### __How to set priorities of Edge modules__
+
+Here's an example of how to set the priorities of IoT Edge modules through Az CLI:
+
+Create a deployment manifest `deployment.json` JSON file that has your IoT Edge deployment specification. More information about the IoT Edge deployment manifest can be found [here][1].
```
Comment (kgremban) [inline doc/ModulePriority.md]: As a rule of thumb, "here" is discouraged as link alt-text. For accessibility reasons, it's hard for people with visual impairment to use because customers who use screen readers often scan link alt-text without context of the surrounding sentences. For non-visually impaired readers, it doesn't set good expectations for what they'll see on the other end of that link. For writers, its difficult to maintain especially in the case when a link might break.

Recommended practice is to use the title of the target page.

File: doc/ModulePriority.md (hunk 1 above)
Comment (shantanu1singh) [reply doc/ModulePriority.md]: You're right, I somehow missed removing this.

File: doc/ModulePriority.md (hunk 3 above)
Comment (shantanu1singh) [reply doc/ModulePriority.md]: Ah I see. Sure, I'll go ahead and update this.

File: doc/ModulePriority.md (hunk 2 above)
Comment (shantanu1singh) [reply doc/ModulePriority.md]: I wasn’t are aware of the deployment prioritization design in Edge and I’d gone ahead and created a design spec (and subsequently implemented it) using the priority order starting from zero which is the highest priority.

The pull-request implementing this feature has been merged with the `master` branch of IoT Edge and I have notified our customers on how to use it with the priorities beginning from zero. Do we still wish to change this implementation to be more aligned with the deployment prioritization scheme?

File: doc/DbStoreSizeLimits.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
+# How to configure database store size limits of Edge Hub
+
+By default, the IoT Edge system module Edge Hub stores data in an on device disk-based database store. This database store is used to store configuration data and messages.
+
+Support for an in-memory database store is also provided for customers who wish to limit the amount of disk I/O on the device and can be enabled by setting the `UsePersistentStorage` environment variable of the module to `false`.
```
Comment (varunpuranik) [inline doc/DbStoreSizeLimits.md]: I would reword this as - Users can configure the EdgeHub to use an in-memory store instead, by setting the UsePersistentStorage environment to 'false'.

File: doc/DbStoreSizeLimits.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+The maximum size of the in-memory database store can be controlled by configuring the limit in the desired properties of the module.
```
Comment (varunpuranik) [inline doc/DbStoreSizeLimits.md]: Maybe we can remove this line.

File: doc/DbStoreSizeLimits.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+## __Use case__
+
+Customers who wish to use the in-memory database store of Edge Hub and also wish to control the maximum size that the database can grow to, will find this feature useful.
+
+## __Configuration__
+
+To enable a max size limit of the in-memory database stores, please set the `MaxSizeBytes` property in the `StoreLimits` configuration section, of the `StoreAndForwardConfiguration` desired property element of Edge Hub, to the maximum size in bytes that you wish to limit the database store size to.
+There are some other configuration changes that need to be made to enable this feature successfully:
+
+### __Required configuration__
```
Comment (varunpuranik) [inline doc/DbStoreSizeLimits.md]: This is duplicated on line #9 above

File: doc/DbStoreSizeLimits.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+```JSON
+"storeAndForwardConfiguration": {
+    "timeToLiveSecs": 7200,
+    "storeLimits": {
+        "maxSizeBytes": 1000000
+    }
+}
+```
+
+To deploy the deployment.json file to your device, you can use the Az CLI ([How to install][2]) with the Az CLI IoT Extension.
```
Comment (varunpuranik) [inline doc/DbStoreSizeLimits.md]: I am not sure if all this should be in this doc. 
@kgremban ?

File: doc/InMemoryDbBackupAndRestore.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
+# In-memory database backup and restore of IoT Edge system modules
```
Comment (varunpuranik) [inline doc/InMemoryDbBackupAndRestore.md]: Should we instead have one doc that talks about IoT Edge storage configuration?

File: doc/ModulePriority.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
+# How to configure priorities for modules
```
Comment (varunpuranik) [inline doc/ModulePriority.md]: Suggested text: "Module startup order priorities". 
@kgremban  to confirm

File: doc/ModulePriority.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+Modules that possess the same priority will be created at the same time and will have no deterministic startup order imposed amongst themselves. Modules with higher priority values will be created and started first and once Edge Agent has made an attempt to create and start them (only the modules with a desired state of `Running` will be started), it will proceed with the creation (and optionally startup) of other lower priority modules in order.
+
+**It must be noted that the Edge Agent module will not support the `priority` property and will by default have the highest priority. As is the behavior currently, Edge Agent will always be the first module to get started.**
+
+Modules that do not have a `priority` value specified will be started in a non-deterministic order and will be assigned a priority of 4294967295 which indicates the lowest possible priority. Edge Agent will trigger the creation of these modules after all modules with a higher priority have been created.
+
+**It must be noted that module boot ordering will initially not be supported for the Kubernetes execution mode of IoT Edge.**
```
Comment (varunpuranik) [inline doc/ModulePriority.md]: Nit: Can we be consistent in calling it startup order?
Also suggested text: Please note that Kubernetes mode of IoT Edge does not support module startup order priorities.

File: doc/ModulePriority.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+* The `$edgeHub` module has been assigned a `priority` value of 0.
+* The `SimulatedTemperatureSensor` module has been assigned a `priority` value of 1.
+* The `filtermodule` module has not been assigned any `priority` value which means that it will by default assume the priority of 4294967295 which is the lowest priority value possible.
+
+When this deployment manifest is deployed to a device that does not have any modules running, `$edgeHub` is the first module that will come up followed by the `SimulatedTemperatureSensor` module and then the `filtermodule`.
+
+To deploy the deployment.json file to your device, you can use the Az CLI ([How to install][2]) with the Az CLI IoT Extension.
+
+How to install the Az CLI IoT extension:
```
Comment (varunpuranik) [inline doc/ModulePriority.md]: Again, I don't know if we need this in every doc.

File: doc/DbStoreSizeLimits.md (hunk 7 above)
Comment (kgremban) [reply doc/DbStoreSizeLimits.md]: It's probably not necessary, because I assume a customer reading this has already deployed at least the tempsensor to a device. You could link out to the docs ([Deploy Azure IoT Edge modules with Azure CLI](https://docs.microsoft.com/azure/iot-edge/how-to-deploy-modules-cli)) to provide guidance instead.

File: doc/ModulePriority.md (hunk 9 above)
Comment (kgremban) [reply doc/ModulePriority.md]: "How to configure..." is good because having a verb in the title helps readers know what they'll be doing. But I agree with the later comment about being consistent around terminology, and startup order priority is more clear. So maybe "How to configure module startup order" or "How to configure startup order priority for modules"

File: doc/DbStoreSizeLimits.md (hunk 7 above)
Comment (shantanu1singh) [reply doc/DbStoreSizeLimits.md]: Going to change it as per your recommendation

File: doc/ModulePriority.md (hunk 9 above)
Comment (shantanu1singh) [reply doc/ModulePriority.md]: Going to change it as per your recommendation

File: doc/ModulePriority.md (hunk 11 above)
Comment (shantanu1singh) [reply doc/ModulePriority.md]: Changing it to refer to the tutorial recommended by @kgremban

File: doc/ModulePriority.md (hunk 10 above)
Comment (shantanu1singh) [reply doc/ModulePriority.md]: Done

File: doc/InMemoryDbBackupAndRestore.md (hunk 8 above)
Comment (shantanu1singh) [reply doc/InMemoryDbBackupAndRestore.md]: I initially had that thought but I wasn't sure what all should be documented with respect to storage configuration (in terms of details and implementation) so I went ahead with this instead.

We can add that document later and refer to it in these other docs.
